### PR TITLE
[MM-31250] Implement export command

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR implements a new `export` command to create, list, download and delete server generated bulk export files.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-31250

#### Related PRs

https://github.com/mattermost/mattermost-server/pull/16633

### CODE REVIEW

File: vendor/modules.txt [hunk 1]
```diff
@@ -237,13 +239,13 @@ github.com/mattermost/go-i18n/i18n/language
...
 github.com/mattermost/gorp
-# github.com/mattermost/ldap v0.0.0-20191128190019-9f62ba4b8d4d
+# github.com/mattermost/ldap v0.0.0-20201202150706-ee0e6284187d
 github.com/mattermost/ldap
 # github.com/mattermost/logr v1.0.13
 github.com/mattermost/logr
 github.com/mattermost/logr/format
 github.com/mattermost/logr/target
-# github.com/mattermost/mattermost-server/v5 v5.3.2-0.20201204150807-d137cf0aa5b2
+# github.com/mattermost/mattermost-server/v5 v5.3.2-0.20210105114834-d357b20d2253
```
Comment (streamer45) [inline vendor/modules.txt:248]: As usual, will update this once the server PR gets merged.

File: commands/export_e2e_test.go [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+	importFilePath := filepath.Join(serverPath, "tests", importName)
+	exportPath, err := filepath.Abs(filepath.Join(*s.th.App.Config().FileSettings.Directory,
+		*s.th.App.Config().ExportSettings.Directory))
+	s.Require().Nil(err)
+
+	s.Run("no permissions", func() {
+		printer.Clean()
+
+		err := exportListCmdF(s.th.Client, &cobra.Command{}, nil)
+		s.Require().NotNil(err)
```
Comment (isacikgoz) [inline commands/export_e2e_test.go:34]: nittest nit:
```suggestion
		s.Require().NoError(err)
```

File: commands/export_e2e_test.go [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+
+		downloadPath, err := filepath.Abs(exportName)
+		s.Require().Nil(err)
+		err = utils.CopyFile(importFilePath, downloadPath)
+		s.Require().Nil(err)
+		defer os.Remove(downloadPath)
+
+		err = exportDownloadCmdF(c, cmd, []string{exportName, downloadPath})
+		s.Require().NotNil(err)
+		s.Require().Equal("export file already exists", err.Error())
```
Comment (isacikgoz) [inline commands/export_e2e_test.go:218]:
```suggestion
		s.Require().EqualError(err, "export file already exists")
```

**Review by isacikgoz — APPROVED** [review]
Looks great @streamer45 🚀 Have some minor suggestions to be considered.

File: commands/export.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+var ExportJobCmd = &cobra.Command{
+	Use:   "job",
+	Short: "List and show export jobs",
+}
+
+var ExportJobListCmd = &cobra.Command{
+	Use:     "list",
+	Example: " export job list",
```
Comment (mgdelacroix) [inline commands/export.go:61]: nit: for consistency, examples should have two spaces before the text

File: commands/export.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	Use:   "create",
+	Short: "Create export file",
+	Args:  cobra.NoArgs,
+	RunE:  withClient(exportCreateCmdF),
+}
+
+var ExportDownloadCmd = &cobra.Command{
+	Use:     "download [exportname] [filepath]",
+	Short:   "Download export files",
+	Example: " export download export_file.zip",
```
Comment (mgdelacroix) [inline commands/export.go:33]: Can we add a couple of examples following the new format here? I was a bit confused with the command at first. Maybe something like:

```suggestion
	Example: `  # you can indicate the name of the export and it's destination path
  $ mmctl export download samplename sample_export.zip
  
  # or if you only indicate the name, the path would match it
  $ mmctl export download sample_export.zip`,
```

File: commands/export_e2e_test.go [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+		err := exportListCmdF(c, &cobra.Command{}, nil)
+		s.Require().Nil(err)
+		s.Require().Len(printer.GetLines(), 1)
+		s.Require().Empty(printer.GetErrorLines())
+		s.Equal("No export files found", printer.GetLines()[0])
+	})
+
+	s.RunForSystemAdminAndLocal("some exports", func(c client.Client) {
+		cmd := &cobra.Command{}
+		if c == s.th.LocalClient {
```
Comment (mgdelacroix) [inline commands/export_e2e_test.go:52]: Why do we need to set this flag explicitly for local tests?

**Review by mgdelacroix — APPROVED** [review]
LGTM! :+1: Added a couple of nonblocking comments

File: commands/export_e2e_test.go (hunk 6 above)
Comment (streamer45) [reply commands/export_e2e_test.go:52]: We don't. It was an imports only quirk. Removing :+1:

**Comment by streamer45** [comment]
@metanerd 

Some help would be great. I've tried to bump Golang version for the Gitlab pipeline but it's still complaining.

**Review by metanerd — APPROVED** [review]
Looks great!

**Review by lindy65 — APPROVED** [review]
Thanks @streamer45 - this will be tested after merge.  

I have created test cases for this PR and have linked the test cases to the Jira ticket.  Please use the Zephyr (TM4J) code of the tests and link them to the E2E tests for these commands.  If you need assistance please chat to @mgdelacroix 

Thanks!

File: commands/export_e2e_test.go [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+		s.Require().Empty(printer.GetErrorLines())
+
+		actual, err := ioutil.ReadFile(downloadPath)
+		s.Require().Nil(err)
+
+		s.Require().Equal(expected, actual)
+	})
+}
+
+func (s *MmctlE2ETestSuite) TestExportJobShow() {
```
Comment (mgdelacroix) [inline commands/export_e2e_test.go:318]:
```suggestion
func (s *MmctlE2ETestSuite) TestExportJobShowCmdF() {
```

File: commands/export_e2e_test.go [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+
+		err := exportJobShowCmdF(c, &cobra.Command{}, []string{job.Id})
+		s.Require().Nil(err)
+		s.Require().Empty(printer.GetErrorLines())
+		s.Require().Len(printer.GetLines(), 1)
+		s.Require().Equal(job, printer.GetLines()[0].(*model.Job))
+	})
+}
+
+func (s *MmctlE2ETestSuite) TestExportJobList() {
```
Comment (mgdelacroix) [inline commands/export_e2e_test.go:355]:
```suggestion
func (s *MmctlE2ETestSuite) TestExportJobListCmdF() {
```

**Comment by amyblais** [comment]
@streamer45 Is this fine to go in to v5.34? The related server PR is included in tomorrow's cloud release + v5.33, but this mmctl PR would go in to v5.34 when it's merged into master.

**Comment by streamer45** [comment]
> @streamer45 Is this fine to go in to v5.34? The related server PR is included in tomorrow's cloud release + v5.33, but this mmctl PR would go in to v5.34 when it's merged into master.

It's not a problem from my side. It's this one that depends on the server one so it's fine to get this in later.

**Pull request merged**